File: swarmmanagercontroller/src/main/java/co/uk/swarmbit/docker/api/configs/ConfigsApiImpl.java
package co.uk.swarmbit.docker.api.configs;

import co.uk.swarmbit.docker.api.common.AbstractApiImpl;
import co.uk.swarmbit.docker.api.configs.parameters.ConfigsListParameters;
import co.uk.swarmbit.docker.api.common.json.ConfigCreateResponseJson;
import co.uk.swarmbit.docker.api.common.json.ConfigJson;
import co.uk.swarmbit.docker.api.configs.parameters.ConfigsCreateParameters;
import co.uk.swarmbit.docker.api.configs.parameters.ConfigsUpdateParameters;
import co.uk.swarmbit.rest.RestResponseType;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class ConfigsApiImpl extends AbstractApiImpl implements ConfigsApi {

    private static final String CONFIGS_PATH = "/configs";

    private static final String CREATE_PATH = "/create";

    private static final String UPDATE_PATH = "/update";

    @Override
    public List<ConfigJson> listConfigs(String swarmId, ConfigsListParameters parameters) {
        return listObjects(CONFIGS_PATH, swarmId, new RestResponseType<List<ConfigJson>>() {}, parameters);
    }

    @Override
    public ConfigJson inspectConfig(String swarmId, String id) {
        return inspectObject(CONFIGS_PATH, swarmId, new RestResponseType<ConfigJson>() {}, id);
    }

    @Override
    public ConfigCreateResponseJson createConfig(String swarmId, ConfigsCreateParameters parameters) {
        return createObject(CONFIGS_PATH + CREATE_PATH, swarmId, new RestResponseType<ConfigCreateResponseJson>() {}, parameters);
    }

    @Override
    public void deleteConfig(String swarmId, String id) {
        deleteObject(CONFIGS_PATH + "/" + id, swarmId, new RestResponseType<Void>() {});
    }

    @Override
    public void updateConfig(String swarmId, String id, ConfigsUpdateParameters parameters) {
        updateObject(CONFIGS_PATH + "/" + id + UPDATE_PATH, swarmId, new RestResponseType<Void>() {}, parameters, parameters);
    }
}

File: swarmmanagercontroller-new/src/main/kotlin/com/swarmbit/docker/api/configs/ConfigsApiImpl.kt
package com.swarmbit.docker.api.configs

import com.swarmbit.docker.api.common.AbstractApiImpl
import com.swarmbit.docker.api.common.client.DockerWebClient
import com.swarmbit.docker.api.common.rest.model.ResponseType
import com.swarmbit.docker.api.configs.model.Config
import com.swarmbit.docker.api.configs.model.ConfigCreateResponse
import com.swarmbit.docker.api.configs.parameters.ConfigsCreateParameters
import com.swarmbit.docker.api.configs.parameters.ConfigsListParameters
import com.swarmbit.docker.api.configs.parameters.ConfigsUpdateParameters
import io.micronaut.context.annotation.Context

@Context
class ConfigsApiImpl(dockerWebClient: DockerWebClient) : AbstractApiImpl(dockerWebClient), ConfigsApi {

    companion object {
        private const val CONFIGS_PATH = "/configs"
        private const val CREATE_PATH = "/create"
        private const val UPDATE_PATH = "/update"
    }

    override fun listConfigs(swarmId: String, parameters: ConfigsListParameters): List<Config> {
        return listObjects(CONFIGS_PATH, swarmId, object : ResponseType<List<Config>>() {}, parameters)
    }

    override fun inspectConfig(swarmId: String, id: String): Config {
        return inspectObject(CONFIGS_PATH, swarmId, object : ResponseType<Config>() {}, id)
            ?: throw IllegalArgumentException("No config found for swarmId ($swarmId) and configId ($id)")
    }

    override fun createConfig(swarmId: String, parameters: ConfigsCreateParameters): ConfigCreateResponse {
        return createObject(CONFIGS_PATH + CREATE_PATH, swarmId, object : ResponseType<ConfigCreateResponse>() {}, parameters)!!
    }

    override fun deleteConfig(swarmId: String, id: String) {
        deleteObject<Void>("$CONFIGS_PATH/$id", swarmId, object : ResponseType<Void>() {})
    }

    override fun updateConfig(swarmId: String, id: String, parameters: ConfigsUpdateParameters) {
        updateObject("$CONFIGS_PATH/$id$UPDATE_PATH", swarmId, object : ResponseType<Void>() {}, parameters, parameters)
    }
}
